Add Printer Support: Brother DCP-L8410CDW

shahab-vahedi opened the pull request:
The LPR drivers and CUPS wrappers from Debian packages provided on [Brother's Support page](https://support.brother.com/g/b/producttop.aspx?c=eu_ot&lang=en&prod=dcpl8410cdw_eu).

#### Testing the changes
- I tested the changes in this PR: **YES**

#### New package
- This new package conforms to the [package requirements](https://github.com/void-linux/void-packages/blob/master/CONTRIBUTING.md#package-requirements): **YES**

#### Local build testing
- I built this PR locally for my native architecture, `x86_64-glibc`

shahab-vahedi commented:
Ping!

shahab-vahedi commented:
Ping^2!

shahab-vahedi commented:
Ping^3!

shahab-vahedi commented:
Ping^4!

shahab-vahedi commented:
Could someone take a look at these? I've been using them for a month and see no surprises.

ahesford reviewed (changes requested):
I'm not inclined to accept these packages. At best, this ought to be considered for inclusion as a restricted package; not because the EULA prohibits repackaging and distribution but, barring maintainership by a core team member that is willing to assume responsibility, repacking deb archives provided in a corporate download site can be a fragile operation.

Notwithstanding the general objection, there are a number of quality issues:
- The EULAs and wrapper script should not be vendored. Fetch them from the deb archive or the web at build time.
- The README.voidlinux documents do not belong. These documents are intended for instructions specific to Void as a consequence of packaging decisions. The contents of your files are general usage instructions that users are expected to seek from the upstream vendor.
- The INSTALL and REMOVE scripts are inappropriate and should be removed. Our packages don't restart users' services just by virtue of installation or removal. Users can restart their own services when appropriate.

shahab-vahedi commented:
> I'm not inclined to accept these packages. At best, this ought to be considered for inclusion as a restricted package

I can add `restricted=yes` to the `template`. HOWEVER, if the community doesn't want this package, then I keep it to myself. I just wanted to share it with other Void users to save them some time.

> The EULAs and wrapper script should not be vendored...
For the EULA, my point of reference was the [agree.html](https://github.com/void-linux/void-packages/commits/master/srcpkgs/brother-brscan4/files/agree.html) already existing in `bother-brscan4` package.

Regarding the _wrapper script_, If you're referring to `cupswrapperdcpl8410cdw.void`, it's a trimmed down version of the original script in the `*.deb` file to suit a Void Linux system. The alternative would be to have the original one installed and apply a _patch_ on top of it. For ease of maintenance, I chose the former approach. For the record, the [brother-dcp197c-cupswarpper](https://github.com/void-linux/void-packages/blob/master/srcpkgs/brother-dcp197c-cupswrapper/files/cupswrapper.void), as it is in `master` branch of 24-Oct-2024, was my inspiration.

> ...Fetch them from the deb archive or the web at build time.

I find nothing in the `*.deb` file. As for the _web_, I neither am sure where that would be nor have I the will to locate it. Maybe I should just remove the `agree.html`.

> The INSTALL and REMOVE scripts are inappropriate and should be removed.
> Our packages don't restart users' services just by virtue of installation or removal.
> Users can restart their own services when appropriate.

This was a mimic of the `*.deb` package itself, but sure, I can remove them. Keep in mind, one of them, `brother-dcp8410-lpr/INSTALL`, is creating a directory and has nothing to do with the services.

I appreciate that someone has finally taken a peek at this. Whatever the outcome of this discussion will be, it will inherently be the same for my other [PR](https://github.com/void-linux/void-packages/pull/51636) which upgrades (and fixes) the (broken) existing package of [brother-brscan4](https://github.com/void-linux/void-packages/commits/master/srcpkgs/brother-brscan4/).

ahesford commented:
Vendoring the wrapper script entirely just obscures the changes you made, which makes it harder to maintain as the upstream version drifts over time. It also makes your changes harder to review, because we can't see if the patch is appropriate. Use the version from the deb archive and include a patch that makes the necessary changes for the Void package.

We provide a `make_dirs` variable to allow directories to me made during package installation. That should be preferred over a custom `INSTALL` script to do the same.

If the `brother-brscan4` package venders a EULA, it is doing the wrong thing. The license should be captured at package build time because it is subject to change between the moment you fetched a copy and the moment the package fetched distfiles and built a package. If it isn't in the deb archive, you can grab it at build time from wherever you got it in the first place.

shahab-vahedi commented:
> Vendoring the wrapper script entirely just obscures the changes you made, which makes it harder to maintain as the upstream version drifts over time. It also makes your changes harder to review, because we can't see if the patch is appropriate. Use the version from the deb archive and include a patch that makes the necessary changes for the Void package.

Ack.

> We provide a `make_dirs` variable to allow directories to me made during package installation. That should be preferred over a custom `INSTALL` script to do the same.

Ack.

> If the `brother-brscan4` package venders a EULA, it is doing the wrong thing. The license should be captured at package build time because it is subject to change between the moment you fetched a copy and the moment the package fetched distfiles and built a package. If it isn't in the deb archive, you can grab it at build time from wherever you got it in the first place.

Could you name a package that already does something like that? so I can make myself familiar with the whole picture/mechanism.

shahab-vahedi commented:
> > If the `brother-brscan4` package venders a EULA, it is doing the wrong thing. The license should be captured at package build time because it is subject to change between the moment you fetched a copy and the moment the package fetched distfiles and built a package. If it isn't in the deb archive, you can grab it at build time from wherever you got it in the first place.
> 
> Could you name a package that already does something like that? so I can make myself familiar with the whole picture/mechanism.

The `post_extract()` in `brother-brscan3` package does that exactly.

shahab-vahedi commented:
@ahesford 
ChangeLog for the new changes:

brother-dcp8410-lpr
-------------------
- Removed `README.voidlinux`
  As a consequence, `INSTALL` and `REMOVE` scripts are added to
  perform the actions mentioned in there.

- Added a mechanism to get an up-to-date ToS from vendor's website.
  Also verify if the checksum of the agreement has remained the same.

- Did NOT mark the package as restricted, because ToS allows redistribution.

brother-dcp8410-cups
--------------------
- Removed `INSTALL` and `REMOVE` scripts

- Renamed `README.void` to `README.maintainer`
  I really hate to see this doc go away. It is not documented at
  the vendor website. I've gathered those notes while studying the
  wrapper script.

- Replaced the `cupswrapperdcpl8410cdw.void` script with a patch
  The patch is 328 lines while the drop-in script was only 152 lines.
  I believe it is easier to wrap one's head around the drop-in wrapper
  script (pun intended) rather than going through the 300+ lines of patch.
  Specially that the upstream package hasn't been updated for years, there
  isn't virtually any maintenance work to update the drop-in wrapper.

- Added a mechanism to get an up-to-date ToS from vendor's website.
  Also verify if the checksum of the agreement has remained the same.

- Did NOT mark the package as restricted, because ToS allows redistribution.

ahesford commented:
I think you misinterpreted my comments:
- Marking the package restricted is a precondition for consideration, without consideration for terms of service that may allow redistribution. It should be restricted because the build process is fragile. A precompiled deb offered by a corporate sponsor may disappear at any time, or may break and be unfixable becuase of other Void updates. We do not need to accept the possibility that officially repacking this software may break our repository at an inopportune moment.
- The `README.maintainer` does not belong in our repository. As I noted, users are expected to familiarize themselves with the maintenance of printers in CUPS should they wish to use it. Void is not a repository for general-purpose documentation.
- The INSTALL and REMOVE scripts remain inappropriate. Void does not alter general system configuration just by virtue of package installation or removal. Removing or adding printers when this package is installed or removed is not appropriate; users may wish to install or remove packages without rewriting printcap. This is especially true because the `-lpr` package is a dependency of the `-cups` package, but CUPS tends to own `/etc/printcap`; other packages shouldn't be modifying that file at all.

In addition:
- In the `-lpr` package, why are you rearranging the executable structure? If the package installs executables in `x86_64` and `i686` subdirectories, you should leave them where they are, but remove the unnecessary arch for each of the `i686` and `x86_64` packages.
- The wrapper script in `-cups` is poorly written (it doesn't even spell `lpdwrapper` correctly) and isn't really suitable for Void at all. The script should **never** copy files into the `/usr/share` or `/usr/lib*` trees, because those are owned by the package manager. The attempt to start, stop or restart services is also poorly conceived. The Void template should just symlink the PPD file(s) as appropriate, so that the package manager can manage those links.
- There is no need for `curl` or fetching the license file at runtime; you can just add the EULA URL to the `distfiles` array in each case, using the following construct:
    ```sh
    distfiles="<regular-distfile>
     https://support.brother.com/g/s/agreement/English_lpr/agree.html>eula-${version}.html"
    skip_extraction="eula-${version}.html"

    post_extract() {
        sed -n -e <...> "${XBPS_SRCDISTDIR}/${pkgname}-${version}/eula-${version}.html" > LICENSE
    }
    ```
